youtube_dlc/YoutubeDL.py

```diff
@@ -1875,17 +1889,21 @@ def dl(name, info):
                         postprocessors = [merger]
 
                     def compatible_formats(formats):
-                        video, audio = formats
+                        # TODO: some formats actually allow this (mkv, webm, ogg, mp4), but not all of them.
+                        video_formats = [format for format in formats if format.get('vcodec') != 'none']
+                        audio_formats = [format for format in formats if format.get('acodec') != 'none']
+                        if len(video_formats) > 2 or len(audio_formats) > 2:
+                            return False
+
                         # Check extension
-                        video_ext, audio_ext = video.get('ext'), audio.get('ext')
-                        if video_ext and audio_ext:
-                            COMPATIBLE_EXTS = (
-                                ('mp3', 'mp4', 'm4a', 'm4p', 'm4b', 'm4r', 'm4v', 'ismv', 'isma'),
-                                ('webm')
-                            )
-                            for exts in COMPATIBLE_EXTS:
-                                if video_ext in exts and audio_ext in exts:
-                                    return True
+                        exts = set(format.get('ext') for format in formats)
+                        COMPATIBLE_EXTS = (
+                            set(('mp3', 'mp4', 'm4a', 'm4p', 'm4b', 'm4r', 'm4v', 'ismv', 'isma')),
+                            set(('webm',)),
```

Comment:
Maybe I don't understand this correctly but is there a downside to add mkv to this as well? Or is it not needed since mkv is a fall back extension anyway?

Reply:
This is a rather unlikely scenario, but as I read the code now, this might trigger Matroska audio and WebM video being multiplexed into a WebM container; in general this is not possible, as WebM is a proper subset of Matroska that supports only a handful of codecs.

In my opinion this function should probably be completely re-written. One could have a table of container formats: for each container a list of codecs it is capable of containing, whether it supports multiple streams of a single type, probably a preference value as well. The optimal container would be picked on that basis. One problem is that the codec of a stream is not always known, only the container (that is, the extension).

For the time being, I just mostly kept what was there before.